go/login/login.go

```diff
@@ -0,0 +1,306 @@
+)
+
+// Contains the <message> information
+type Message struct {
+	HostIdType string
+	HostId     string
+	Action     string
+}
+
+// Constructs a Message according to the structure: [<hostid-type>:]<hostid>[/<action>].
+func (msgInfo *Message) Construct() []byte {
+	msg := ""
+	if msgInfo.HostId != "" {
+		msg += escape(msgInfo.HostIdType) + ":"
+	}
+	msg += fmt.Sprintf("%s", escape(msgInfo.HostId))
+
+	if msgInfo.Action != "" {
+		msg += fmt.Sprintf("/%s", msgInfo.Action)
+	}
+	return []byte(msg)
+}
+
+// Escapes the string so it can be safely placed inside a URL path segment,
+// replacing "/#?" special characters and not replacing "!*'();:@&=+$,[]" special characters.
+func escape(s string) string {
+	res := url.PathEscape(s)
+	for _, c := range "!*'();:@&=+$,[]" {
+		st := string(c)
+		strings.ReplaceAll(res, url.PathEscape(st), st)
+	}
+	return res
+}
+
+// Contains the <glome-handshake> information.
+type Handshake struct {
+	Prefix       byte
+	UserKey      glome.PublicKey
+	MsgTagPrefix []byte
+}
+
+// Contains the challenge information.
+type URLResponse struct {
+	V             byte
+	HandshakeInfo Handshake
+	Msg           Message
+	d             *glome.Dialog
+}
+
+func NewResponse(serviceKeyId uint8, serviceKey glome.PublicKey, userKey glome.PrivateKey,
+	V byte, hostIdType string, hostId string, action string) (*URLResponse, error) {
+	var prefix byte
+	var r URLResponse
+
+	d, e := userKey.Exchange(&serviceKey)
+	if e != nil {
+		return nil, e
+	}
+	r.d = d
+
+	r.Msg = Message{hostIdType, hostId, action}
+
+	if serviceKeyId == 0 {
+		// If no key ID was specified, send the first key byte as the ID.
+		prefix = serviceKey[0] & 0x7f
+	} else {
+		prefix = serviceKeyId & 0x7f
+	}
+	userPublic, e := userKey.Public()
+	if e != nil {
+		return nil, e
+	}
+	HandshakeInfo := Handshake{prefix, *userPublic, r.GetTag()}
+	r.HandshakeInfo = HandshakeInfo
+
+	return &r, nil
+}
+
+// Validates if the received tag corresponding to the base64-url encoded message constructed from the Message.
+func (r *URLResponse) ValidateAuthCode(tag string) bool {
+	return r.d.CheckEncoded(tag, r.Msg.Construct(), 0)
+}
+
+// Returns the tag corresponding to the Msg.
+func (r *URLResponse) GetTag() []byte {
+	return r.d.Tag(r.Msg.Construct(), 0)
+}
+
+// Returns a base64-encoded response token.
+func (r *URLResponse) GetEncToken() string {
+	return base64.URLEncoding.EncodeToString(r.GetTag())
+}
+
+// Client side login-lib handler.
+type Client struct {
+	ServiceKey   glome.PublicKey
+	UserKey      glome.PrivateKey
+	ServiceKeyId uint8
+	response     *URLResponse
+}
+
+// Client constructor. Sets Client.ServiceKey, Client.UserKey, Client.ServiceKeyId to the corresponding values
+// and Client.response to nil.
+func NewClient(sk glome.PublicKey, uk glome.PrivateKey, sId uint8) *Client {
+	return &Client{sk, uk, sId, nil}
+}
+
+func (c *Client) Construct(V byte, hostIdType string, hostId string, action string) (string, error) {
```

Comment:
Please document what is constructed. I'd prefer if both the client and the server struct would be operating on the same type (`*url.URL` in this case). However, this would require knowing about the scheme and host, which is also inconvenient.

Reply:
Updated to `string` usage instead of `*url.URL`. Had thought about both variant, but firstly decided the second variant as it contains scheme which is sometimes in the url.

Reply:
Makes sense, but please update the documentation because we're not adding a `url-prefix` here.

Reply:
Fixed.